configuration/config2/EnvironmentNode.cpp

```diff
@@ -0,0 +1,385 @@
+        // If this value must be unique, make sure it is
+        if (attrIt->second->getSchemaValue()->isUniqueValue())
+        {
+            bool found = false;
+            std::vector<std::string> allValues = attrIt->second->getAllValues();
+            std::set<std::string> unquieValues;
+            for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
+            {
+                auto ret = unquieValues.insert(*it);
+                found = ret.second;
+            }
+
+            if (found)
+            {
+                status.addUniqueStatusMsg(statusMsg::error, m_id, attrIt->second->getName(), "", "Attribute value must be unique");
+            }
+        }
+
+        //
+        // Does this value need to be from another set of values?
+        if (attrIt->second->getSchemaValue()->isFromUniqueValueSet())
+        {
+            bool found = false;
+            std::vector<std::string> allValues = attrIt->second->getSchemaValue()->getAllKeyRefValues();
+            for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
+                found = *it == attrIt->second->getValue();
+            if (!found)
+            {
+                status.addStatusMsg(statusMsg::error, m_id, attrIt->second->getName(), "", "Attribute value must be from a unique set");
+            }
+        }
+    }
+
+    //
+    // Now check all children
+    if (includeChildren)
+    {
+        for (auto childIt = m_children.begin(); childIt != m_children.end(); ++childIt)
+        {
+            childIt->second->validate(status);
+        }
+    }
+}
+
+
+std::vector<std::string> EnvironmentNode::getAllAttributeValues(const std::string &attrName) const
+{
+    std::vector<std::string> values;
+    std::shared_ptr<EnvironmentNode> pParentNode = m_pParent.lock();
+    if (pParentNode)
+    {
+        std::vector<std::shared_ptr<EnvironmentNode>> nodes = pParentNode->getChildren(m_name);
+        for (auto it = nodes.begin(); it != nodes.end(); ++it)
+        {
+            values.push_back((*it)->getAttributeValue(attrName));
+        }
+    }
+    return values;
+}
+
+
+const std::shared_ptr<EnvironmentValue> EnvironmentNode::getAttribute(const std::string &name) const
+{
+    std::shared_ptr<EnvironmentValue> pValue;
+    auto it = m_attributes.find(name);
+    if (it != m_attributes.end())
+    {
+        pValue = it->second;
+    }
+    return pValue;
+}
+
+
+std::vector<std::shared_ptr<SchemaItem>> EnvironmentNode::getInsertableItems() const
+{
+    std::vector<std::shared_ptr<SchemaItem>> insertableItems;
+    std::map<std::string, unsigned> childCounts;
+
+    //
+    // Iterate over the children and for each, create a childCount entry based on the
+    // child node's configuration type
+    for (auto childIt = m_children.begin(); childIt != m_children.end(); ++childIt)
+    {
+        std::string itemType = childIt->second->getSchemaItem()->getItemType();
+        auto findIt = childCounts.find(itemType);
```

Comment:
auto is a preferred way to retrieve iterators for STL containers. It takes into account the container type, the type of elements, and constness of the container itself. Since iterators are used in a common way for containers, it simplifies the code as well. Also, should some aspect of the container change, auto ensures that all iterators are correct w/o having to inspect and change each use.

As far as pre-incrementing, it can be more efficient than post increment. The -> binds first, so the value is incremented for the case listed.

I added a comment to the pre-increment line in question to give it semantic meaning.

Reply:
The advantages of auto are well understood, in fact, I wouldn't be surprised if I've recommended you to use it in the past. You have to be aware of the drawbacks as well, and be judicious. In this case, it seems alright, but the variable name could have been much better if expanded to findIterator, or even better childIterator.